optimizely/decision_service.py

```diff
@@ -323,128 +329,216 @@ def get_variation(
+
+        Args:
+          config: Instance of ProjectConfig.
+          flag_key: Key of the flag.
+          rule: Experiment rule.
+          user: ID and attributes for user.
+          options: Decide options.
+
+        Returns:
+          Decision namedtuple consisting of experiment and variation for the user and
+          array of log messages representing decision making.
+        """
+        decide_reasons = []
+
+        # check forced decision first
+        optimizely_decision_context = OptimizelyUserContext.OptimizelyDecisionContext(flag_key, rule.key)
+
+        forced_decision_variation, reasons_received = user.find_validated_forced_decision(
+            optimizely_decision_context,
+            options)
+        decide_reasons += reasons_received
+
+        if forced_decision_variation:
+            return forced_decision_variation, decide_reasons
+
+        # regular decision
+        decision_variation, variation_reasons = self.get_variation(config, rule, user, options)
+        decide_reasons += variation_reasons
+        return decision_variation, decide_reasons
+
+    def get_variation_from_delivery_rule(self, config, feature, rules, rule_index, user, options):
+        """ Checks for delivery rule if decision is forced and returns it.
+            Otherwise returns a regular decision.
+
+        Args:
+          config: Instance of ProjectConfig.
+          flag_key: Key of the flag.
+          rules: Experiment rule.
+          user: ID and attributes for user.
+          options: Decide options.
+
+        Returns:
+          If forced decision, it returns namedtuple consisting of forced_decision_variation and skip_to_everyone_else
+          and decision reason log messages.
+
+          If regular decision it returns a tuple of bucketed_variation and skip_to_everyone_else
+          and decision reason log messages
+        """
+        decide_reasons = []
+        skip_to_everyone_else = False
+        bucketed_variation = None
+
+        # check forced decision first
+        rule = rules[rule_index]
+        optimizely_decision_context = OptimizelyUserContext.OptimizelyDecisionContext(feature.key, rule.key)
+        forced_decision_variation, reasons_received = user.find_validated_forced_decision(optimizely_decision_context,
+                                                                                          options)
+
+        decide_reasons += reasons_received
+
+        if forced_decision_variation:
+            return (forced_decision_variation, skip_to_everyone_else), decide_reasons
+
+        # regular decision
+        user_id = user.user_id
+        attributes = user.get_user_attributes()
+        bucketing_id = self._get_bucketing_id(user_id, attributes)
+
+        everyone_else = (rule_index == len(rules) - 1)
+        logging_key = "Everyone Else" if everyone_else else str(rule_index + 1)
+
+        rollout_rule = config.get_experiment_from_id(rule.id)
```

Comment:
We already have 'rule' in line 449. Remove this line.

Reply:
Done